### PR TITLE
CacheContent header

### PR DESCRIPTION
**Pull request by hiroya8649**
對c++極度生疏
請再多指教

### CODE REVIEW

File: src/include/CacheContent.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#define CACHE_CONTENT_H_
+
+#include <string>
+
+/**
+ * 包裝實際的檔案
+ * 讓其具有過期以及讀檔狀態的概念
+ */
+
+class CacheContent
```
Comment (john0312) [inline src/include/CacheContent.hpp]: Please add class comment.

File: src/include/CacheContent.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+{
+public:
+    CacheContent();
```
Comment (john0312) [inline src/include/CacheContent.hpp]: Is this class copyable?

File: src/include/CacheContent.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    ~CacheContent();
+    std::string *getContent();
```
Comment (john0312) [inline src/include/CacheContent.hpp]: Why do you return a pointer to std::string? Do you expect the caller to modify its content?
If you want to avoid copying the string, you can return a const reference.

Comment (john0312) [reply src/include/CacheContent.hpp]: Please add class comment.

File: src/include/CacheContent.hpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    bool isValid(); // isReady and is not expired
+    bool isPending(); // not isReady and is not expired
+    bool isExpired(); // readAt is too old
+private:
+    std::string *content; // the content, the worker need to read the file and put it here.
+    bool isReady; // if the worker done read, call the callback and set isReady to true.
```
Comment (john0312) [inline src/include/CacheContent.hpp]: Please use underscore. `is_ready`
https://google.github.io/styleguide/cppguide.html#Variable_Names

File: src/include/CacheManager.hpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
+/**
```
Comment (john0312) [inline src/include/CacheManager.hpp]: Please separate into a different commit/PR.

File: src/include/CacheManager.hpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ * 將檔案內容載入記憶體
+ * 內部保存一個map，將key(應為request url)對應到一個CacheContent
+ */
+class CacheManager // singleton
+{
+public:
+    CacheManager();
+    ~CacheManager();
+    CacheManager &getInstance();
```
Comment (john0312) [inline src/include/CacheManager.hpp]: Usually we return the raw pointer for non-const return type.

File: src/include/CacheManager.hpp [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    // 取得map內某個key的內容
+    // 如果該key不存在，建立一個新的CacheContent
+    // 如果該key存在但其CacheContent太舊，reload
+    // 如果該key存在但尚未ready，則其尚在讀檔，這段不確定該如何處理讓正在讀檔的worker知道讀檔完須多call一隻callback
+    // 如果該key存在且ready，回傳其內容
+    void get(std::string key, std::function<void(CacheContent &)> callback);
```
Comment (john0312) [inline src/include/CacheManager.hpp]: const ref for input argument.

Comment (john0312) [reply src/include/CacheManager.hpp]: Pass const CacheContent& into the callback if you don't want to allow the callback to modify the cache. Otherwise, pass in a pointer.

Comment (hiroya8649) [reply src/include/CacheManager.hpp]: fixed

File: src/include/CacheManager.hpp (hunk 5 above)
Comment (hiroya8649) [reply src/include/CacheManager.hpp]: 有拆開來commit

File: src/include/CacheContent.hpp (hunk 4 above)
Comment (hiroya8649) [reply src/include/CacheContent.hpp]: fixed

File: src/include/CacheContent.hpp (hunk 2 above)
Comment (hiroya8649) [reply src/include/CacheContent.hpp]: fixed

File: src/include/CacheManager.hpp (hunk 6 above)
Comment (hiroya8649) [reply src/include/CacheManager.hpp]: 可以請教是什麼意思嗎？

Comment (john0312) [reply src/include/CacheManager.hpp]: 如果要回傳一個可以被改(non-const)的變數，一般我們使用指標。如果要回傳一個不可以被改的變數，一般我們使用constant reference。

**Comment by john0312** [comment]
請把你們commit squash起來，謝謝。

**Comment by hiroya8649** [comment]
> 請把你們commit squash起來，謝謝。

@john0312 
請問是要squash成什麼樣子呢？
我的經驗大部分都是在pr通過要merge時才做squash
謝謝

File: src/include/CacheManager.hpp (hunk 6 above)
Comment (hiroya8649) [reply src/include/CacheManager.hpp]: fixed

**Comment by john0312** [comment]
Squash成最終要merge的樣子。以這個PR來說，一個commit放CacheContent，一個commit放CacheManager。

File: src/include/CacheContent.hpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ * }
+ *   
+ */
+
+class CacheContent
+{
+public:
+    CacheContent();
+    CacheContent();
+    CacheContent(const CacheContent& other) = default;
```
Comment (john0312) [inline src/include/CacheContent.hpp]: Please comment that it's not copyable.

File: src/include/CacheContent.hpp [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    CacheContent& operator=(const CacheContent&) = default;
+    ~CacheContent();
+    const std::string getContent();
```
Comment (john0312) [inline src/include/CacheContent.hpp]: Please add comment for method.

File: src/include/CacheManager.hpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef CACHE_MANAGER_H_
+#define CACHE_MANAGER_H_
+
+#include <string>
```
Comment (john0312) [inline src/include/CacheManager.hpp]: Please order the include.
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/include/CacheManager.hpp [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+ * cm.get("some/file.txt", someCallback);
+ */
+class CacheManager // singleton
+{
+public:
+    CacheManager();
+    CacheManager(const CacheManager& other) = delete;
+    CacheManager& operator=(const CacheManager&) = delete;
+    ~CacheManager();
+    const CacheManager &getInstance();
```
Comment (john0312) [inline src/include/CacheManager.hpp]: Return a pointer?

Comment (john0312) [reply src/include/CacheManager.hpp]: If you return a const reference, caller will not be able to call non-const method.

File: src/include/CacheManager.hpp [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    // 取得map內某個key的內容
+    // 如果該key不存在，建立一個新的CacheContent
+    // 如果該key存在但其CacheContent太舊，reload
+    // 如果該key存在但尚未ready，則其尚在讀檔，這段不確定該如何處理讓正在讀檔的worker知道讀檔完須多call一隻callback
+    // 如果該key存在且ready，回傳其內容
+    void get(const std::string key, const std::function<void(const CacheContent &)> callback);
```
Comment (john0312) [inline src/include/CacheManager.hpp]: const reference for input.

File: src/include/CacheManager.hpp [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+private:
+    std::map<std::string, CacheContent *> hash;
```
Comment (john0312) [inline src/include/CacheManager.hpp]: Please add comment for variable.

File: src/include/CacheManager.hpp [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+ * Example: 
+ * CacheManager cm = new CacheManager();
+ * cm.get("some/file.txt", someCallback);
+ */
+class CacheManager // singleton
+{
+public:
+    CacheManager();
+    CacheManager(const CacheManager& other) = delete;
+    CacheManager& operator=(const CacheManager&) = delete;
```
Comment (john0312) [inline src/include/CacheManager.hpp]: Add comment that it's not copyable.

**Comment by john0312** [comment]
Please clang-format and fix the git commits by rebasing/squashing.

File: src/include/CacheContent.hpp (hunk 8 above)
Comment (hiroya8649) [reply src/include/CacheContent.hpp]: fixed

File: src/include/CacheContent.hpp (hunk 9 above)
Comment (hiroya8649) [reply src/include/CacheContent.hpp]: fixed

File: src/include/CacheManager.hpp (hunk 10 above)
Comment (hiroya8649) [reply src/include/CacheManager.hpp]: fixed

File: src/include/CacheManager.hpp (hunk 11 above)
Comment (hiroya8649) [reply src/include/CacheManager.hpp]: fixed

File: src/include/CacheManager.hpp (hunk 12 above)
Comment (hiroya8649) [reply src/include/CacheManager.hpp]: fixed

File: src/include/CacheManager.hpp (hunk 13 above)
Comment (hiroya8649) [reply src/include/CacheManager.hpp]: fixed

File: src/include/CacheManager.hpp (hunk 14 above)
Comment (hiroya8649) [reply src/include/CacheManager.hpp]: fixed
